### ISSUE TITLE
Require 2FA enrollment for all new users

### ISSUE THREAD

**Issue by ewdurbin** [pypi/warehouse#13762]
**What's the problem this feature will solve?**

Since we will require it for all uploads beginning in 2024, we should add 2FA enrollment for all new users as soon as possible.

**Describe the solution you'd like**

Add 2FA enrollment to routine new user registration flow.

**Comment by miketheman** [comment]
## User Registration experience today:

1. Visit registration page
2. Complete registration signup form
3. User is now logged in, and an email has been sent for verification

At this point, there's nothing preventing the user from **never** verifying their email address, other than wanting to publish a package - Create API Token requires a verified email.

4. User verifies email, they can create an API token, and also start 2FA enrollment

## The question is

At what step of the flow should we **require** the user to enroll in 2FA?

- Should we perform a similar check as email verification for whether 2FA is enrolled prior to creating an API token?
- Should we require a verified email address/2FA before allowing logged-in actions?

Basically, something about the user registration flow has to change, and today we allow unverified registrations to continue to use the logged-in aspects of PyPI - manage collaborators, yank/delete releases, setup trusted publishing, etc.

**Comment by di** [comment]
> Should we require a verified email address/2FA before allowing logged-in actions?

I think this makes sense.

**Comment by ristomcgehee** [comment]
How does this sound for a new user registration flow?

Instead of having a single view for registering a user, the user goes through these 5 steps, each in its own view:
1. Basic Info - Enters name, email, username, and password. This view will be just like the current registration view, except it will say "Continue" or "Next" instead of "Create Account".
2. Choose 2FA Method - Inform the user that new user registration requires setting up 2FA. Include the link to the help page about 2FA. Have the two buttons: "Add 2FA with authentication application" and "Add 2FA with security device (e.g. USB key)".
3. Set Up 2FA - Show either the TOTP or webauthn configuration page, based on what they picked on the previous step.
4. Display Recovery Codes - Basically the same as the current recovery codes page: Explain the reason for the recovery codes, show the 10 codes, have copy and save buttons.
5. Burn Recovery Code - Same as current page, maybe change button to say "Verify & Create Account".

Then the user will be taken to the manage account page where they'll be more or less forced to verify their email address (nice improvement, Mike).

I know this ordering is different than what existing users go through, but I think if we were to do recovery codes first, we would have to explain that were going to require them to set up 2FA on the next step and also explain that they need recovery codes, too. Instead we can just have them set up 2FA and then next tell them they need recovery codes in case they lose their 2FA. 

If it works technically, I think it might make sense to not create the user until after step \#5 and have the POST to create the user include the recovery codes, TOTP secret/Webauthn public, etc.

**Comment by miketheman** [comment]
https://github.com/pypi/warehouse/pull/14126 implemented the requirement for having a verified primary email for management actions other than account management.

**Comment by ewdurbin** [comment]
Thank you @miketheman !
